packages/react-pages/src/node/dynamic-modules/PageStrategy.ts

```diff
@@ -0,0 +1,297 @@
+    findPages(pagesDir, helpers)
+
+    this.getPages = () => pagesPromise
+    this.close = () => watchers.forEach((w) => w.close())
+
+    function addPageData({
+      pageId,
+      key = 'main',
+      dataPath,
+      staticData,
+    }: PageData) {
+      if (!pageId.startsWith('/'))
+        throw Error(
+          `addPageData error: pageId should start with "/", but got "${pageId}"`
+        )
+
+      const existingPage = pageCache[pageId]
+      const page =
+        existingPage ||
+        (pageCache[pageId] = {
+          data: {},
+          staticData: {},
+        })
+
+      if (dataPath) {
+        if (page.data[key])
+          throw Error(
+            `addPageData conflict: data with key "${key}" already exists`
+          )
+        page.data[key] = dataPath
+      }
+
+      if (staticData) {
+        if (page.staticData[key])
+          throw Error(
+            `addPageData conflict: staticData with key "${key}" already exists`
+          )
+        page.staticData[key] = staticData
+      }
+
+      if (currentFile) {
+        let dataKeys = currentFile.dataKeys.get(pageId)
+        if (!dataKeys) currentFile.dataKeys.set(pageId, (dataKeys = []))
+        dataKeys.push(key)
+      }
+
+      if (existingPage) {
+        changedPages.add(pageId)
+        emitChange()
+      }
+    }
+
+    function removePageData(dataKeys: string[], pageId: string) {
+      const page = pageCache[pageId]
+
+      dataKeys.forEach((key) => {
+        delete page.data[key]
+        delete page.staticData[key]
+      })
+
+      if (hasKeys(page.data) || hasKeys(page.staticData)) {
+        changedPages.add(pageId)
+        emitChange()
+      } else {
+        delete pageCache[pageId]
+      }
+    }
+
+    function hasKeys(obj: object) {
+      for (const _ in obj) return true
+      return false
+    }
+
+    function watchFiles(
+      baseDir: string,
+      arg2?: string | string[] | FileHandler,
+      arg3?: FileHandler
+    ) {
+      let globs: string[]
+      let handler: FileHandler
+
+      if (typeof arg2 === 'function') {
+        globs = ['**/*']
+        handler = arg2
+      } else {
+        globs = Array.isArray(arg2) ? arg2 : [arg2 || '**/*']
+        handler = arg3 || defaultPageLoader
+      }
+
+      // Strip trailing slash and make absolute
+      baseDir = path.resolve(pagesDir, baseDir)
+
+      const processFile = async (
+        type: 'add' | 'change' | 'unlink',
+        file: File
+      ) => {
+        // Clear any data associated with this file.
+        file.dataKeys.forEach(removePageData)
+
+        if (type === 'unlink') {
+          delete fileCache[file.path]
+        } else {
+          if (type === 'change') {
+            file.content = null
+            file.dataKeys.clear()
+          }
+          currentFile = file
+          try {
+            const pageData = await handler(file)
+            if (pageData) {
+              addPageData(pageData)
```

Comment:
Should we call `addPageData` for users? What if user want to associate this file with multiple pages? User can call it by `helper.addPageData`.

Reply:
It's optional. The handler can return nothing and call `addPageData` manually. ;)